File: C++/Problems/Problems/problem_12.cpp
#include "problem_12.h"

#include <iostream>
#include <ostream>

bool problem_12::test() {
    auto num = 3749;

    std::string expected = "MMMDCCXLIX";

    auto result = intToRoman(num);

    std::cout << result << std::endl;

    return expected == result;
}

std::string problem_12::intToRoman(int num) {
    auto numString = std::to_string(num);
    std::string result = "";

    if (numString.size() == 4)
    {
        result += std::string(numString[0] - '0', 'M');
        numString = numString.substr(1);
    }


    if (numString.size() == 3)
    {
        appendStuff(result, 'C', 'D', 'M', numString[0]);
        numString = numString.substr(1);
    }

    if (numString.size() == 2)
    {
        appendStuff(result, 'X', 'L', 'C', numString[0]);
        numString = numString.substr(1);
    }

    if (numString.size() == 1)
    {
        appendStuff(result, 'I', 'V', 'X', numString[0]);
        numString = numString.substr(1);
    }


    return result;
}

void problem_12::appendStuff(std::string &result, char one, char five, char dec, char src) {
    if (src < '4')
        result += std::string(src - '0', one);
    else if (src == '4') {
        result.push_back(one);
        result.push_back(five);
    }
    else if (src < '9') {
        result.push_back(five);
        result += std::string(src - '0' - 5, one);
    }
    else {
        result.push_back(one);
        result.push_back(dec);
    }
}

File: Python/Problems/Problems/problem12.py
from .problem import Problem

class Problem12(Problem):
    def test(self):
        num = 3749

        expected = "MMMDCCXLIX"

        result = self.intToRoman(num)

        print(result)

        return expected == result

    def intToRoman(self, num):
        numString = str(num)
        result = ""

        if len(numString) == 4:
            result += 'M' * (ord(numString[0]) - ord('0'))
            numString = numString[1::]


        if len(numString) == 3:
            result = self.appendStuff(result, 'C', 'D', 'M', numString[0])
            numString = numString[1:]

        if len(numString) == 2:
            result = self.appendStuff(result, 'X', 'L', 'C', numString[0])
            numString = numString[1:]

        if len(numString) == 1:
            result = self.appendStuff(result, 'I', 'V', 'X', numString[0])
            numString = numString[1:]

        return result

    def appendStuff(self, result, one, five, dec, src):
        if ord(src) < ord('4'):
            result += one * (ord(src) - ord('0'))
        elif ord(src) == ord('4'):
            result += one + five
        elif ord(src) < ord('9'):
            result += five + one * (ord(src) - ord('0') - 5)
        else:
            result += one + dec
        return result